src/wallet/wallet2.cpp

```diff
@@ -3873,36 +4351,334 @@ bool wallet2::load_tx(const std::string &signed_filename, std::vector<tools::wal
+
+    THROW_WALLET_EXCEPTION_IF(get_transaction_prefix_hash (tx) != get_transaction_prefix_hash(ptx.tx),
+        error::wallet_internal_error, "Transaction prefix does not match data");
+
+    // Tests passed, sign
+    std::vector<unsigned int> indices;
+    for (const auto &source: sources)
+      indices.push_back(source.real_output);
+
+    // work out the permutation done on sources
+    std::vector<size_t> ins_order;
+    for (size_t n = 0; n < sd.sources.size(); ++n)
+    {
+      for (size_t idx = 0; idx < sources.size(); ++idx)
+      {
+        THROW_WALLET_EXCEPTION_IF((size_t)sources[idx].real_output >= sources[idx].outputs.size(),
+            error::wallet_internal_error, "Invalid real_output");
+        if (sources[idx].outputs[sources[idx].real_output].second.dest == sd.sources[n].outputs[sd.sources[n].real_output].second.dest)
+          ins_order.push_back(idx);
+      }
+    }
+    THROW_WALLET_EXCEPTION_IF(ins_order.size() != sources.size(), error::wallet_internal_error, "Failed to work out sources permutation");
+
+    for (auto &sig: ptx.multisig_sigs)
+    {
+      if (sig.ignore != local_signer)
+      {
+        ptx.tx.rct_signatures = sig.sigs;
+
+        rct::keyV k;
+        for (size_t idx: sd.selected_transfers)
+          k.push_back(get_multisig_k(idx, sig.used_L));
+        tools::apply_permutation(ins_order, k);
+
+        rct::key skey = rct::zero();
+        for (const auto &msk: get_account().get_multisig_keys())
+        {
+          crypto::public_key pmsk = get_multisig_signing_public_key(msk);
+
+          if (sig.signing_keys.find(pmsk) == sig.signing_keys.end())
+          {
+            sc_add(skey.bytes, skey.bytes, rct::sk2rct(msk).bytes);
+            sig.signing_keys.insert(pmsk);
+          }
+        }
+        THROW_WALLET_EXCEPTION_IF(!rct::signMultisig(ptx.tx.rct_signatures, indices, k, sig.msout, skey),
+            error::wallet_internal_error, "Failed signing, transaction likely malformed");
+
+        sig.sigs = ptx.tx.rct_signatures;
+      }
+    }
+
+    const bool is_last = exported_txs.m_signers.size() + 1 >= m_multisig_threshold;
+    if (is_last)
+    {
+      if (m_multisig)
+      {
+        // when the last signature on a multisig tx is made, we select the right
+        // signature to plug into the final tx
+        bool found = false;
+        for (const auto &sig: ptx.multisig_sigs)
+        {
+          if (sig.ignore != local_signer && exported_txs.m_signers.find(sig.ignore) == exported_txs.m_signers.end())
+          {
+            THROW_WALLET_EXCEPTION_IF(found, error::wallet_internal_error, "More than one transaction is final");
+            ptx.tx.rct_signatures = sig.sigs;
+            found = true;
+          }
+        }
+        THROW_WALLET_EXCEPTION_IF(!found, error::wallet_internal_error, "Final signed transaction not found");
+      }
+      const crypto::hash txid = get_transaction_hash(ptx.tx);
+      if (store_tx_info())
```

Comment:
Isn't this already done inside `load_multisig_tx_from_file `?

Reply:
In sign_multisig_tx. Might be technically unneeed most of the time, though the committer might not have signed it, if it's a N-1/N.

Reply:
I mean this line in `load_multisig_tx`:
```c++
if (store_tx_info())
```
https://github.com/monero-project/monero/pull/2134/files#diff-9d580668dab930045943902048e2ac22R4480

This is called from `load_multisig_tx_from_file` which is called from `sign_multisig_tx_from_file` which is before `sign_multisig_tx` is called. I believe it's redundant since I don't see any code path that reaches `sign_multisig_tx` without going through `load_multisig_tx`.

Reply:
Ah, I'd forgot I'd done that. Still, I see no good reason to remove it from the other place, it's a tiny amount of code and work for something which makes it less prone to bugs should other code change.

Reply:
I agree.